src/main/java/is/citizen/sdk/crypto/CitizenCrypto.java

```diff
@@ -0,0 +1,364 @@
+            ECPublicKey publicKey = (ECPublicKey) keyFactory.generatePublic(publicKeySpec);
+
+            Signature verificationFunction = Signature.getInstance(AUTH_SIGNATURE_ALGORITHM);
+            verificationFunction.initVerify(publicKey);
+            verificationFunction.update(data);
+
+             if (verificationFunction.verify(signedData)) {
+                 return true;
+             }
+        } catch (NoSuchAlgorithmException | InvalidKeySpecException | InvalidKeyException |
+                 SignatureException e)
+        {
+            throw new CryptoException("Unable to verify signature: " + e.getMessage());
+        }
+
+        return false;
+    }
+
+    public Optional<Map> encryptMap(Map map, String publicKey)
+        throws CryptoException {
+        Map<String, String> encmap = new HashMap<>();
+
+        if ((map != null) && map.keySet().size() > 0) {
+            Iterator i = map.entrySet().iterator();
+            while (i.hasNext()) {
+                Map.Entry<String, String> pair = (Map.Entry) i.next();
+
+                // decrypted key
+                Optional<String> decproptype = Optional.ofNullable(pair.getKey());
+
+                // decrypted val
+                Optional<String> decval = Optional.ofNullable(pair.getValue());
+                if (decval.isPresent()) {
+                    Optional<String> encval = encryptString(decval.get(), publicKey);
+                    if (encval.isPresent()) {
+                        encmap.put(decproptype.get(), encval.get());
+                    }
+                }
+            }
+        }
+
+        return Optional.ofNullable(encmap);
+    }
+
+    public static Optional<Map> decryptMap(Map map, String privKey, String secret)
+        throws CryptoException {
+        Map decmap = new HashMap<>();
+        if (map != null) {
+            Iterator i = map.keySet().iterator();
+            while (i.hasNext()) {
+                String key = (String) i.next();
+                Optional<String> encval = Optional.ofNullable((String) map.get(key));
+                if (encval.isPresent()) {
+                    Optional<String> decval = decryptString(encval.get(), privKey, secret);
+                    if (decval.isPresent()) {
+                        decmap.put(key, decval.get());
+                    }
+                }
+            }
+        }
+
+        return Optional.ofNullable(decmap);
+    }
+
+    public static Optional<String> encryptString(String inputString, String publicKey)
+        throws CryptoException {
+        String encryptedText = null;
+
+        try {
+            if (StringUtils.hasText(inputString)) {
+                byte[] cipherText;
+                cipherText = encrypt(inputString.getBytes("UTF8"), getPublicKeyFromEncodedString(publicKey));
+                encryptedText = Base64.getEncoder().encodeToString(cipherText);
+            }
+        } catch (Exception e) {
+            throw new CryptoException(e.getMessage());
```

Comment:
Better exception messages. Also,  we should avoid just catching Exception. We should be more specific

Reply:
Fixed.